packages/fixie/src/auth.ts

```diff
@@ -0,0 +1,207 @@
+  const validKeys = ['apiUrl', 'apiKey'];
+  const invalidKeys = Object.keys(config).filter((key) => !validKeys.includes(key));
+  for (const key of invalidKeys) {
+    term('❓ Ignoring invalid key ').yellow(key)(` in ${fullPath}\n`);
+  }
+  return config as FixieConfig;
+}
+
+/** Save the client configuration to the given file. */
+export function SaveConfig(config: FixieConfig, configFile: string) {
+  const fullPath = untildify(configFile);
+  const dirName = path.dirname(fullPath);
+  if (!fs.existsSync(dirName)) {
+    fs.mkdirSync(dirName, { recursive: true });
+  }
+  if (fs.existsSync(fullPath)) {
+    // Merge the new config with the existing config, so we don't
+    // overwrite any fields that are not specified.
+    const currentConfig = yaml.load(fs.readFileSync(fullPath, 'utf8')) as object;
+    const mergedConfig = { ...currentConfig, ...config };
+    fs.writeFileSync(fullPath, yaml.dump(mergedConfig));
+  } else {
+    fs.writeFileSync(fullPath, yaml.dump(config));
+  }
+}
+
+/** Returns an authenticated FixieClient, or null if the user is not authenticated. */
+export async function Authenticate({
+  apiUrl,
+  configFile,
+}: {
+  apiUrl?: string;
+  configFile?: string;
+}): Promise<FixieClient | null> {
+  // The precedence for selecting the API URL and key is:
+  //   1. apiUrl argument to this function. (The key cannot be passed as an argument.)
+  //   2. FIXIE_API_URL and FIXIE_API_KEY environment variables.
+  //   3. apiUrl and apiKey fields in the config file.
+  //   4. Fallback value for apiUrl (constant defined above).
+  const config = LoadConfig(configFile ?? FIXIE_CONFIG_FILE);
+  const useApiUrl = apiUrl ?? process.env.FIXIE_API_URL ?? config.apiUrl ?? FIXIE_API_URL;
+  const useApiKey = process.env.FIXIE_API_KEY ?? config.apiKey;
+  if (!useApiKey) {
+    // No key available. Need to punt.
+    return null;
+  }
+  const client = FixieClient.Create(useApiUrl, useApiKey);
+  const userInfo = await client.userInfo();
+  if (userInfo.is_anonymous) {
+    return null;
+  }
+  return client;
+}
+
+/** Returns an authenticated FixieClient, starting an OAuth flow to authenticate the user if necessary. */
+export async function AuthenticateOrLogin({
+  apiUrl,
+  configFile,
+  forceReauth,
+}: {
+  apiUrl?: string;
+  configFile?: string;
+  forceReauth?: boolean;
+}): Promise<FixieClient> {
+  if (!forceReauth) {
+    const client = await Authenticate({
+      apiUrl,
+      configFile,
+    });
+    if (client) {
+      const userInfo = await client.userInfo();
+      if (!userInfo.is_anonymous) {
+        return client;
+      }
+    }
+  }
+
+  const apiKey = await oauthFlow(apiUrl ?? FIXIE_API_URL);
+  const config: FixieConfig = {
+    apiUrl: apiUrl ?? FIXIE_API_URL,
+    apiKey,
+  };
+  SaveConfig(config, configFile ?? FIXIE_CONFIG_FILE);
+  const client = await Authenticate({ apiUrl, configFile: configFile ?? FIXIE_CONFIG_FILE });
+  if (!client) {
+    throw new Error('Failed to authenticate');
```

Comment:
Can we provide a more actionable error message? Should the dev retry? Ask for help in Discord? Is it possibly their fault?